File: android/MicroService/src/main/java/org/dweb_browser/microservice/ipc/helper/IpcEvent.kt
package org.dweb_browser.microservice.ipc.helper

import com.google.gson.*
import com.google.gson.annotations.JsonAdapter
import kotlinx.serialization.Serializable
import org.dweb_browser.helper.ProxySerializer
import org.dweb_browser.helper.toBase64
import org.dweb_browser.helper.toUtf8
import java.lang.reflect.Type


@Serializable
data class IpcEventJsonAble(
  val name: String, val data: String, val encoding: IPC_DATA_ENCODING
) : IpcMessage(IPC_MESSAGE_TYPE.EVENT) {
  fun toIpcEvent() = IpcEvent(name, data, encoding)
}

object IpcEventSerializer :
  ProxySerializer<IpcEvent, IpcEventJsonAble>(IpcEventJsonAble.serializer(),
    { jsonAble },
    { toIpcEvent() })

@Serializable(with = IpcEventSerializer::class)
@JsonAdapter(IpcEvent::class)
class IpcEvent(
  val name: String, val data: Any /*String or ByteArray*/, val encoding: IPC_DATA_ENCODING
) : IpcMessage(IPC_MESSAGE_TYPE.EVENT), JsonSerializer<IpcEvent>, JsonDeserializer<IpcEvent> {

  companion object {
    fun fromBinary(name: String, data: ByteArray) = IpcEvent(name, data, IPC_DATA_ENCODING.BINARY)

    fun fromBase64(name: String, data: ByteArray) =
      IpcEvent(name, data.toBase64(), IPC_DATA_ENCODING.BASE64)

    fun fromUtf8(name: String, data: ByteArray) = fromUtf8(name, data.toUtf8())

    fun fromUtf8(name: String, data: String) = IpcEvent(name, data, IPC_DATA_ENCODING.UTF8)
  }

  val binary by lazy {
    dataToBinary(data, encoding)
  }

  val text by lazy {
    dataToText(data, encoding)
  }

  val jsonAble by lazy {
    when (encoding) {
      IPC_DATA_ENCODING.BINARY -> fromBase64(
        name,
        (data as ByteArray),
      )

      else -> this
    }.let {
      IpcEventJsonAble(
        name,
        data as String,
        encoding,
      )
    }
  }

  override fun serialize(
    src: IpcEvent, typeOfSrc: Type, context: JsonSerializationContext
  ) = JsonObject().also { jsonObject ->
    with(src.jsonAble) {
      jsonObject.add("type", context.serialize(type))
      jsonObject.addProperty("name", name)
      jsonObject.addProperty("data", data as String)
      jsonObject.add("encoding", context.serialize(encoding))
    }
  }

  override fun deserialize(
    json: JsonElement, typeOfT: Type, context: JsonDeserializationContext
  ): IpcEvent = json.asJsonObject.let { obj ->
    IpcEvent(
      name = obj["name"].asString,
      data = obj["data"].asString,
      encoding = context.deserialize(obj["encoding"], IPC_DATA_ENCODING::class.java)
    )
  }

}

File: desktop-dev/src/core/ipc/ipc-message/IpcEvent.ts
import { simpleDecoder } from "../../../helper/encoding.ts";
import { $dataToBinary, $dataToText } from "../helper/const.ts";
import { IPC_DATA_ENCODING } from "./internal/IpcData.ts";
import { IPC_MESSAGE_TYPE, ipcMessageBase } from "./internal/IpcMessage.ts";

export type $IpcEvent = ReturnType<typeof _ipcEvent>;

const _ipcEvent = (name: string, data: string | Uint8Array, encoding: IPC_DATA_ENCODING, orderBy?: number) =>
  ({
    ...ipcMessageBase(IPC_MESSAGE_TYPE.EVENT),
    name,
    data,
    encoding,
    orderBy,
  } as const);

export const ipcEvent = Object.assign(_ipcEvent, {
  fromBase64(name: string, data: Uint8Array, orderBy?: number) {
    return ipcEvent(name, simpleDecoder(data, "base64"), IPC_DATA_ENCODING.BASE64, orderBy);
  },
  fromBinary(name: string, data: Uint8Array, orderBy?: number) {
    return ipcEvent(name, data, IPC_DATA_ENCODING.BINARY, orderBy);
  },
  fromUtf8(name: string, data: Uint8Array, orderBy?: number) {
    return ipcEvent(name, simpleDecoder(data, "utf8"), IPC_DATA_ENCODING.UTF8, orderBy);
  },
  fromText(name: string, data: string, orderBy?: number) {
    return ipcEvent(name, data, IPC_DATA_ENCODING.UTF8, orderBy);
  },
  binary(event: $IpcEvent) {
    return $dataToBinary(event.data, event.encoding);
  },
  text(event: $IpcEvent) {
    return $dataToText(event.data, event.encoding);
  },
});
